google/compute_operation.go

```diff
@@ -4,58 +4,64 @@ import (
 	"bytes"
 	"fmt"
 	"log"
+	"strings"
 	"time"
 
 	"github.com/hashicorp/terraform/helper/resource"
+
 	"google.golang.org/api/compute/v1"
 )
 
-// OperationWaitType is an enum specifying what type of operation
-// we're waiting on.
-type ComputeOperationWaitType byte
+type ScopeType uint8
 
 const (
-	ComputeOperationWaitInvalid ComputeOperationWaitType = iota
-	ComputeOperationWaitGlobal
-	ComputeOperationWaitRegion
-	ComputeOperationWaitZone
+	Global ScopeType = iota
+	Region
+	Zone
 )
 
 type ComputeOperationWaiter struct {
 	Service *compute.Service
 	Op      *compute.Operation
 	Project string
-	Region  string
-	Type    ComputeOperationWaitType
-	Zone    string
 }
 
 func (w *ComputeOperationWaiter) RefreshFunc() resource.StateRefreshFunc {
 	return func() (interface{}, string, error) {
 		var op *compute.Operation
 		var err error
 
-		switch w.Type {
-		case ComputeOperationWaitGlobal:
-			op, err = w.Service.GlobalOperations.Get(
-				w.Project, w.Op.Name).Do()
-		case ComputeOperationWaitRegion:
-			op, err = w.Service.RegionOperations.Get(
-				w.Project, w.Region, w.Op.Name).Do()
-		case ComputeOperationWaitZone:
-			op, err = w.Service.ZoneOperations.Get(
-				w.Project, w.Zone, w.Op.Name).Do()
+		// Find the scope type and scope name of an operation
+		var scopeType ScopeType
+		var scope string
+		if w.Op.Zone != "" {
+			scopeType = Zone
+			scopeParts := strings.Split(w.Op.Zone, "/")
+			scope = scopeParts[len(scopeParts)-1]
+		} else if w.Op.Region != "" {
+			scopeType = Region
+			scopeParts := strings.Split(w.Op.Region, "/")
+			scope = scopeParts[len(scopeParts)-1]
+		} else {
+			scopeType = Global
+		}
+
+		switch scopeType {
+		case Global:
+			op, err = w.Service.GlobalOperations.Get(w.Project, w.Op.Name).Do()
+		case Region:
+			op, err = w.Service.RegionOperations.Get(w.Project, scope, w.Op.Name).Do()
+		case Zone:
+			op, err = w.Service.ZoneOperations.Get(w.Project, scope, w.Op.Name).Do()
 		default:
```

Comment:
It doesn't seem like this condition could possibly be hit. What if the `op` assignments were all done in the previous if/else statements?